File: ios/Synergy/Common/Views/Charts/CustomxAxisRender.swift
//
//  CustomxAxisRender.swift
//  Synergy
//
//  Created by Anh Nguyen on 12/5/19.
//  Copyright © 2019 Bitmark Inc. All rights reserved.
//

import Charts

final class CustomxAxisRender: XAxisRendererHorizontalBarChart {
    @objc override func computeSize()
    {
        guard let
            xAxis = self.axis as? XAxis
            else { return }
        
        xAxis.labelWidth = 0
        xAxis.labelHeight = 0
        xAxis.labelRotatedWidth = 0
        xAxis.labelRotatedHeight = 0
    }
    
    override func renderAxisLabels(context: CGContext)
    {
        guard
            let xAxis = self.axis as? XAxis
            else { return }
        
        if !xAxis.isEnabled || !xAxis.isDrawLabelsEnabled
        {
            return
        }
        
        let xoffset = xAxis.xOffset
        
        drawLabels(context: context, pos: viewPortHandler.contentLeft + xoffset, anchor: CGPoint(x: 0.0, y: 0.5))
    }
    
    override func drawLabels(context: CGContext, pos: CGFloat, anchor: CGPoint)
    {
        guard
            let xAxis = self.axis as? XAxis,
            let transformer = self.transformer
            else { return }

        let labelFont = xAxis.labelFont
        let labelTextColor = xAxis.labelTextColor
        let labelRotationAngleRadians = xAxis.labelRotationAngle * .pi / 180

        let centeringEnabled = xAxis.isCenterAxisLabelsEnabled

        // pre allocate to save performance (dont allocate in loop)
        var position = CGPoint(x: 0.0, y: 0.0)

        for i in stride(from: 0, to: xAxis.entryCount, by: 1)
        {
            // only fill x values

            position.x = 0.0

            if centeringEnabled
            {
                position.y = CGFloat(xAxis.centeredEntries[i])
            }
            else
            {
                position.y = CGFloat(xAxis.entries[i])
            }

            transformer.pointValueToPixel(&position)

            if viewPortHandler.isInBoundsY(position.y)
            {
                if let label = xAxis.valueFormatter?.stringForValue(xAxis.entries[i], axis: xAxis)
                {
                    drawLabel(
                        context: context,
                        formattedLabel: label,
                        x: pos - CGFloat(3.0),
                        y: position.y - CGFloat(12.0),
                        attributes: [NSAttributedString.Key.font: labelFont, NSAttributedString.Key.foregroundColor: labelTextColor],
                        anchor: anchor,
                        angleRadians: labelRotationAngleRadians)
                }
            }
        }
    }
}

File: android/app/src/main/java/com/bitmark/fbm/util/view/statistic/CustomXAxisRender.kt
/**
 * SPDX-License-Identifier: ISC
 * Copyright © 2014-2019 Bitmark. All rights reserved.
 * Use of this source code is governed by an ISC
 * license that can be found in the LICENSE file.
 */
package com.bitmark.fbm.util.view.statistic

import android.content.Context
import android.graphics.Canvas
import com.bitmark.fbm.R
import com.bitmark.fbm.util.ext.getDimensionPixelSize
import com.github.mikephil.charting.charts.BarChart
import com.github.mikephil.charting.components.XAxis
import com.github.mikephil.charting.renderer.XAxisRendererHorizontalBarChart
import com.github.mikephil.charting.utils.MPPointF
import com.github.mikephil.charting.utils.Transformer
import com.github.mikephil.charting.utils.ViewPortHandler


class CustomXAxisRender(
    private val context: Context,
    viewPortHandler: ViewPortHandler, xAxis: XAxis,
    trans: Transformer, chart: BarChart
) : XAxisRendererHorizontalBarChart(viewPortHandler, xAxis, trans, chart) {

    override fun computeSize() {
        val xAxis = (mAxis as? XAxis) ?: return
        xAxis.mLabelWidth = 0
        xAxis.mLabelHeight = 0
        xAxis.mLabelRotatedWidth = 0
        xAxis.mLabelRotatedHeight = 0
    }

    override fun renderAxisLabels(c: Canvas?) {
        if (!mXAxis.isEnabled || !mXAxis.isDrawLabelsEnabled)
            return

        val xoffset = mXAxis.xOffset

        mAxisLabelPaint.typeface = mXAxis.typeface
        mAxisLabelPaint.textSize = mXAxis.textSize
        mAxisLabelPaint.color = mXAxis.textColor

        val pointF = MPPointF.getInstance(0f, 0f)

        when {
            mXAxis.position == XAxis.XAxisPosition.TOP           -> {
                pointF.x = 0.0f
                pointF.y = 0.5f
                drawLabels(c, mViewPortHandler.contentRight() + xoffset, pointF)

            }
            mXAxis.position == XAxis.XAxisPosition.TOP_INSIDE    -> {
                pointF.x = 1.0f
                pointF.y = 0.5f
                drawLabels(c, mViewPortHandler.contentRight() - xoffset, pointF)

            }
            mXAxis.position == XAxis.XAxisPosition.BOTTOM        -> {
                pointF.x = 1.0f
                pointF.y = 0.5f
                drawLabels(c, mViewPortHandler.contentLeft() - xoffset, pointF)

            }
            mXAxis.position == XAxis.XAxisPosition.BOTTOM_INSIDE -> {
                pointF.x = 0.0f
                pointF.y = 0.5f
                drawLabels(c, mViewPortHandler.contentLeft() + xoffset, pointF)

            }
            else                                                 -> { // BOTH SIDED
                pointF.x = 0.0f
                pointF.y = 0.5f
                drawLabels(c, mViewPortHandler.contentRight() + xoffset, pointF)
                pointF.x = 1.0f
                pointF.y = 0.5f
                drawLabels(c, mViewPortHandler.contentLeft() - xoffset, pointF)
            }
        }

        MPPointF.recycleInstance(pointF)
    }

    override fun drawLabels(c: Canvas?, pos: Float, anchor: MPPointF?) {
        val labelRotationAngleDegrees = mXAxis.labelRotationAngle
        val centeringEnabled = mXAxis.isCenterAxisLabelsEnabled

        val positions = FloatArray(mXAxis.mEntryCount * 2)

        run {
            var i = 0
            while (i < positions.size) {

                // only fill x values
                if (centeringEnabled) {
                    positions[i + 1] = mXAxis.mCenteredEntries[i / 2]
                } else {
                    positions[i + 1] = mXAxis.mEntries[i / 2]
                }
                i += 2
            }
        }

        mTrans.pointValuesToPixel(positions)

        var i = 0
        while (i < positions.size) {

            val y = positions[i + 1]

            if (mViewPortHandler.isInBoundsY(y)) {

                val label = mXAxis.valueFormatter.getAxisLabel(mXAxis.mEntries[i / 2], mXAxis)
                drawLabel(
                    c,
                    label,
                    pos - context.getDimensionPixelSize(R.dimen.dp_3),
                    y - context.getDimensionPixelSize(R.dimen.dp_10),
                    anchor,
                    labelRotationAngleDegrees
                )
            }
            i += 2
        }
    }
}